---
layout: presentation
title: Organization
permalink: /organization/
---
class: middle center

# {{ page. title }}

---

## Learning goals

* **Primary:** Develop and implement an organizational strategy that facillitates reproducibility
--

* **Secondary:**
  * Evaluate different organizational strategies for their ability to foster reproducibility
  * Integrate organizational principles to fix a chaotic project into a logical structure
  * Interface with a project template to improve the reproducibility of your research

---

## Learning outcomes
* Correct an ongoing project to make it more reproducible
* Apply material from the documentation and organization tutorials to start a new project

---

## Noble paper on file organization

* Discusses far more than organization
* Project organization is a form of documentation and helps make written and scripted documentation easier to maintain
* Provides an outline for the rest of these tutorials
  * Organization
  * Literate programming
  * Scripting
  * Automation
  * Version control

.footnote[[Noble 2009 *PLoS Comp Biol*](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424)]
---

.center[![Noble Figure 1]({{ site.baseurl }}/assets/images/noble-figure-1.png)]

> The core guiding principle is simple: Someone unfamiliar with your project should be able to look at your computer files and understand in detail what you did and why ... Most commonly, however, that "someone" is you.

.footnote[[Noble 2009 *PLoS Comp Biol*](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424)]
---

## Refresher on documentation from Noble

* "Record every operation that you perform"
* "Comment generously"
* "~~Avoid~~[DO NOT] editing intermediate files by hand"
* Use a driver script to centralize all processing and analysis
* "Use relative pathways [relative to the project root]"
* "Make the script restartable"

.footnote[[Noble 2009 *PLoS Comp Biol*](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424)]
---

## Compare file organization from two projects

* Examples
	* [Meadows et al. *Microbiome* study](https://github.com/jfmeadow/Meadow_etal_Surfaces/)
	* [Westcott & Schloss *mSphere* study](https://github.com/SchlossLab/Westcott_OptiClust_mSphere_2017/)
* Questions
	* What do you notice about the structure of the two projects?
  * How long did it taketo find the code for Figure 1?
  * Where is the main document?
	* Where would the data go?

---

## An aside about large files

* GitHub cannot easily handle large datasets
* Each file is limited to 100 MB and a repository is limited to 1 GB
* Generally only post the data files critical to an analysis
* Scripts should document where data goes and how it gets there
* We'll talk more about git/GitHub later

---

## General structure for a microbiome study

```
project
|- README.md        # the top level description of content (this doc)
|- CONTRIBUTING.md  # instructions for how to contribute to your project
|- LICENSE.md       # the license for this project
|- CITATION.md      # instructions on how to cite the project
|
|- submission/
| |- study.Rmd      # executable Rmarkdown for this study, if applicable
| |- study.md       # Markdown (GitHub) version of the *.Rmd file
| |- study.tex      # TeX version of *.Rmd file
| |- study.pdf      # PDF version of *.Rmd file
| |- header.tex     # LaTeX header file to format pdf version of manuscript
| |- references.bib # BibTeX formatted references
| |- XXXX.csl       # csl file to format references for journal XXX
|
|- data             # raw and primary data, are not changed once created
| |- references/    # reference files to be used in analysis
| |- raw/           # raw data, will not be altered
| |- mothur/        # mothur processed data
| +- process/       # cleaned data, will not be altered once created;
|                   # will be committed to repo
|
|- code/            # any code
|
|- results          # all output from workflows and analyses
| |- tables/        # text version of tables to be rendered with kable in R
| |- figures/       # graphs, likely designated for manuscript figures
| +- pictures/      # diagrams, images, and other non-graph graphics
|
|- exploratory/     # exploratory data analysis for study
| |- notebook/      # preliminary analyses
| +- scratch/       # temporary files that can be safely deleted or lost
|
+- Makefile         # executable Makefile for this study, if applicable
```

???

* You are free to do what you want, but keep to the basic priniciples outlined in the Noble paper
* You'll see this again when we discuss the `new_project` repository

---

## Bad project: Introduction

.center[
<iframe width="570" height="320" src="https://www.youtube.com/embed/Fl4L4M8m4d0?ecver=1" frameborder="0" allowfullscreen></iframe>
]

* You have inherited a project from a previous student in the lab
* You need to organize it into a logical structure

---

## Bad project: Exercise

* Download the [data](https://github.com/riffomonas/bad_project/archive/v0.1.zip) to a new directory `~/bad_project`

```{bash}
	wget https://github.com/riffomonas/bad_project/archive/v0.1.zip
	unzip v0.1.zip
	rm v0.1.zip
	mv bad_project-0.1 bad_project
	cd bad_project
```

* Using the recommended directory structure, move files to the appropriate folders
	* Use `mkdir new_directory` to create a new directory
  * Use `mv current_file directory/` to move `current_file` into `directory/`
  * Don't forget the `/` after `directory`

.footnote[Now would be a good time to [log into]() your AWS instance or your local HPC]

---

## General structure for a microbiome study

```
project
|- README.md        # the top level description of content (this doc)
|- CONTRIBUTING.md  # instructions for how to contribute to your project
|- LICENSE.md       # the license for this project
|- CITATION.md      # instructions on how to cite the project
|
|- submission/
| |- study.Rmd      # executable Rmarkdown for this study, if applicable
| |- study.md       # Markdown (GitHub) version of the *.Rmd file
| |- study.tex      # TeX version of *.Rmd file
| |- study.pdf      # PDF version of *.Rmd file
| |- header.tex     # LaTeX header file to format pdf version of manuscript
| |- references.bib # BibTeX formatted references
| |- XXXX.csl       # csl file to format references for journal XXX
|
|- data             # raw and primary data, are not changed once created
| |- references/    # reference files to be used in analysis
| |- raw/           # raw data, will not be altered
| |- mothur/        # mothur processed data
| +- process/       # cleaned data, will not be altered once created;
|                   # will be committed to repo
|
|- code/            # any code
|
|- results          # all output from workflows and analyses
| |- tables/        # text version of tables to be rendered with kable in R
| |- figures/       # graphs, likely designated for manuscript figures
| +- pictures/      # diagrams, images, and other non-graph graphics
|
|- exploratory/     # exploratory data analysis for study
| |- notebook/      # preliminary analyses
| +- scratch/       # temporary files that can be safely deleted or lost
|
+- Makefile         # executable Makefile for this study, if applicable
```

---

## Bad project: Evaluation

* Compare your results to your neighbor
* Once you are done, run `find . | diff - .solution` to compare your results to mine
* What do you notice? Do you agree with my organization?

---

## Using a project template

* The structure of most microbiome analysis projects is fairly standard
* To help with the structure, we created a template, [`new_project`](https://github.com/SchlossLab/new_project/releases/latest)
* Download the zip file for the latest release to your Documents folder

---

## Exercise

.left-column[
* We are going to attempt to generate the paragraph and NMDS plot from [Figure 4](http://aem.asm.org/content/79/17/5112.full) of the Kozich AEM study
* Rename the folder `Kozich_ReAnalysis_AEM_2013`
* We'll set up the project by following along with the instructions in the [`INSTRUCTIONS`](https://github.com/SchlossLab/new_project/blob/master/INSTRUCTIONS.md) file
]

.right-column[
![Mouse analysis section]({{ site.baseurl }}/assets/images/kozich-figure-4.png)
]

---

## Modify README.md

* Open the README.md document in `nano`
* Remove the first line, which starts "Download... "
* Edit the following lines as you see appropriate:
```sh
## TITLE OF YOUR PAPER GOES HERE
YOUR PAPER'S ABSTRACT GOES HERE
```
* We'll leave the rest as it is for now

---

## Changing Filenames

* At the terminal do the following...
```bash
mv newproject_LICENSE.md LICENSE.md
```
* This replaces the license ([CC0](https://creativecommons.org/share-your-work/public-domain/cc0/)) of the `new_project` repository with a license for your analysis ([MIT](https://en.wikipedia.org/wiki/MIT_License))

???

* Go ahead and read and edit the license to include your name and the year
* Note that the MIT license is a permissive license that allows anyone to use your code. This is generally viewed a A Good Thing.
* It would be worth talking with your research group and PI about what license you want to use.
* Don't think you'll be making money off this repository :)
---

## Getting ready for `git`

* At the terminal do the following:
```bash
git init ./
git add .
git commit -m "Initial commit"
```
* This is putting your new project under version control using `git`

---

## Connecting with GitHub

* Go to your home page on GitHub
* Create a new repository as you did before with the paper airplane example
* This time...
  * Give the repository the same name as your project (i.e. `Kozich_ReAnalysis_AEM_2013`).
  * Keep the repository public
  * Don't initialize with a `README` or `.gitignore` files or a license (you've already got those)

---

class: center middle

![Creating a new repository on GitHub]({{ site.baseurl }}/assets/images/github-new-repository-instructions.png)

---

## Connecting with GitHub

* Copy and paste into the terminal the two lines that were highlighted with the red arrow on the previous slide

```bash
Kozich_ReAnalysis_AEM_2013 $ git remote add origin
https://github.com/pschloss/Kozich_ReAnalysis_AEM_2013.git
Kozich_ReAnalysis_AEM_2013 $ git push -u origin master
Counting objects: 26, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (24/24), done.
Writing objects: 100% (26/26), 43.09 KiB | 0 bytes/s, done.
Total 26 (delta 2), reused 0 (delta 0)
remote: Resolving deltas: 100% (2/2), done.
To github.com:pschloss/Kozich_ReAnalysis_AEM_2013.git
 * [new branch]      master -> master
Branch master set up to track remote branch master from origin.
```

* Back in the browser, refresh the screen

---

class: center middle

![New Kozich_ReAnalysis_AEM_2013 repository as shown on GitHub ]({{ site.baseurl }}/assets/images/github-Kozich_ReAnalysis_AEM_2013-repository.png)

---

class: center middle

![Well done - you deserve a high five]({{ site.baseurl }}/assets/images/high-five.gif)

---

## Cleaning up

* Remove the instructions file and commit
```bash
git rm INSTRUCTIONS.md
git commit -m "Remove instructions document"
git push
```
* Refresh your GitHub page and make sure the `INSTRUCTIONS.md` file goes away
* If you ever need the information in the `INSTRUCTIONS.md` file, you can retreive it from the [`new_project` repository](https://github.com/SchlossLab/new_project/blob/master/INSTRUCTIONS.md)

---

## Let's go!

* You need...
  * A partner (preferably someone with mothur experience)
  * Familiarity with [the original paper](http://aem.asm.org/content/79/17/5112.full)
  * The data (hint: grab the data from the "With genomes/metagenomes" link)
  * The reference files
  * mothur
  * Dump the text (for the figure lengend and the paragraph titled, "Scaling up") in the main REAME.md
* Obtain the files and put them in the correct locations
* Do your best to document what you are doing

???

* Learners should be able to download these materials based on what is described in the paper

---

## Helpers

* `wget` - pulls down large data files to the current directory (or `curl -LO`)
* `tar xvf` - unpacks a `.tar` file
* `tar xvzf` - unpacks a `.tar.gz` or `.tgz` file
* `rm` - removes files

---

## When you're done...

* ***Wait*** to commit and push your changes
* Add me/your PI as a collaborator
  * "Settings" -> "Collaborators" -> Enter handle -> "Add collaborator"

---

`data/code/README.md`

```bash
# README

Obtained the Linux version of mothur (v1.39.5) from the mothur GitHub repository
```bash
wget --no-check-certificate https://github.com/mothur/mothur/releases/download/v1.39.5/Mothur.linux_64.zip
unzip Mothur.linux_64.zip
mv mothur code/
rm Mothur.linux_64.zip
rm -rf __MACOSX
`&#8203;``

mothur should output the version by doing...
```bash
code/mothur/mothur -v
`&#8203;``

```

---

`data/raw/README.md`

```bash
# README

Obtained the raw `fastq.gz` files from https://www.mothur.org/MiSeqDevelopmentData.html
&#8203;* Downloaded https://www.mothur.org/MiSeqDevelopmentData/StabilityNoMetaG.tar
&#8203;* Ran the following from the project's root directory

```bash
wget --no-check-certificate https://www.mothur.org/MiSeqDevelopmentData/StabilityNoMetaG.tar
tar xvf StabilityNoMetaG.tar -C data/raw/
rm StabilityNoMetaG.tar
`&#8203;``
```

---

`data/references/README.md`

```bash
# README

Obtained the silva reference alignment from the mothur website:
```bash
wget http://mothur.org/w/images/1/15/Silva.seed_v123.tgz
tar xvzf Silva.seed_v123.tgz silva.seed_v123.align silva.seed_v123.tax
code/mothur/mothur "#get.lineage(fasta=silva.seed_v123.align, taxonomy=silva.seed_v123.tax, taxon=Bacteria);degap.seqs(fasta=silva.seed_v123.pick.align, processors=8)"
mv silva.seed_v123.pick.align data/references/silva.seed.align
rm Silva.seed_v123.tgz silva.seed_v123.*
rm mothur.*.logfile
`&#8203;``

Obtained the RDP reference taxonomy from the mothur website:
```bash
wget -N http://www.mothur.org/w/images/8/88/Trainset14_032015.pds.tgz
tar xvzf Trainset14_032015.pds.tgz trainset14_032015.pds/trainset14_032015.pds.*
mv trainset14_032015.pds/* data/references/
rmdir trainset14_032015.pds
rm Trainset14_032015.pds.tgz
`&#8203;``
```

---

`README.md`

```bash
## Re-analysis of Kozich dataset

&#8203;**Scaling up.** The advantage of the dual-index approach is that a large number of samples
can be sequenced using a number of primers equal to only twice the square root of the
number of samples. To fully evaluate this approach, we resequenced the V4 region of 360
samples that were previously described by sequencing the distal end of the V35 region on
the 454 GS-FLX Titanium platform (18). In that study, we observed a clear separation
between murine fecal samples obtained from 8 C57BL/6 mice at 0 to 9 (early) and 141 to 150
(late) days after weaning, and there was significantly less variation between the late
samples than the early samples. In addition to the mouse fecal samples, we allocated 2
pairs of indices to resequence our mock community. We generated 4.3 million pairs of
sequence reads from the 16S rRNA gene with an average coverage of 9,913 pairs of reads per
sample (95% of the samples had more than 2,454 pairs of sequences) using a new collection
of 8-nt indices (see the supplemental material). Although individual samples were expected
to have various amplification efficiencies, analysis of the number of reads per index did
not suggest a systematic positive or negative amplification bias that could be attributed
to the indices. The combined error rate for the two mock communities was 0.07% before
preclustering and 0.01% after (n = 14,094 sequences). When we used UCHIME to remove
chimeras and rarefied to 5,000 sequences, there was an average of 30.4 OTUs (i.e., 10.4
spurious OTUs). Similar to our previous results, ordination of the mouse fecal samples
again showed the separation between the early and late periods and increased stabilization
with age (Fig. 4) (Mantel test coefficient, 0.81; P < 0.001). These results clearly
indicate that our approach can be scaled to multiplex large numbers of samples.

### Overview

	project
	|- README          # the top level description of content (this doc)
<snip>
```
